x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlanner.java

```diff
@@ -48,8 +55,41 @@ public AssignmentPlanner(List<Node> nodes, List<Model> models) {
     }
 
     public AssignmentPlan computePlan() {
+        return computePlan(true);
+    }
+
+    private AssignmentPlan computePlan(boolean tryAssigningPreviouslyAssignedModels) {
         logger.debug(() -> format("Computing plan for nodes = %s; models = %s", nodes, models));
 
+        AssignmentPlan bestPlan;
+        AssignmentPlan planSatisfyingPreviousAssignments = solveSatisfyingPreviousAssignments();
+        logger.debug(() -> "Plan satisfying previous assignments =\n" + planSatisfyingPreviousAssignments.prettyPrint());
+        if (planSatisfyingPreviousAssignments.arePreviouslyAssignedModelsAssigned() == false && tryAssigningPreviouslyAssignedModels) {
+            AssignmentPlan planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated =
+                solveAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated();
+            logger.debug(
+                () -> "Plan with at least one allocation for previously assigned models =\n"
+                    + planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated.prettyPrint()
+            );
+            if (planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated.arePreviouslyAssignedModelsAssigned()) {
+                bestPlan = planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated;
+            } else {
+                bestPlan = planSatisfyingPreviousAssignments
+                    .countPreviouslyAssignedModelsThatAreStillAssigned() >= planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated
+                        .countPreviouslyAssignedModelsThatAreStillAssigned()
+                            ? planSatisfyingPreviousAssignments
+                            : planAllocatingAtLeastOnceModelsThatWerePreviouslyAllocated;
+            }
+        } else {
+            bestPlan = planSatisfyingPreviousAssignments;
+        }
+
+        logger.debug(() -> "Best plan =\n" + bestPlan.prettyPrint());
+        logger.debug(() -> prettyPrintOverallStats(bestPlan));
+        return bestPlan;
+    }
+
+    private AssignmentPlan solveSatisfyingPreviousAssignments() {
```

Comment:
I think this is for satisfying Current assignments right (figuring out an assignment where currently assigned models still have at least one allocation on a node assignment)? 

These function names are hard to keep together.

Reply:
Correct. I'll try to come up with better naming.